lib/phoenix/socket/transport.ex

```diff
@@ -227,9 +227,17 @@ defmodule Phoenix.Socket.Transport do
     |> do_dispatch(msg, socket)
   end
 
-  defp do_dispatch(nil, %{event: "phx_join", topic: topic} = msg, socket) do
-    if channel = socket.handler.__channel__(topic, socket.transport_name) do
-      socket = %Socket{socket | topic: topic, channel: channel}
+  defp build_channel_socket(%Socket{} = socket, channel, topic) do
+    %Socket{socket |
+            topic: topic,
+            channel: channel,
+            log_join: channel.__channel__(:log_join),
```

Comment:
What if we introduce a private field to the `socket` and store this there? It is going to be counter productive I think to introduce new fields to the socket every time we introduce a new log operation. If we don't want to introduce a generic private field, I would recommend at least grouping the log operations:

    logs: channel.__channel__(:log)

logs will be a map and we should only log if there is a matching key. For example, `%{join: :info}` means only log join with level `:info`.